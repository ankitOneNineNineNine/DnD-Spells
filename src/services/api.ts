/**
 * Apis List
 * @export api
 */
export const api = {
  spells: {
    fetch: "http://www.dnd5eapi.co/api/spells",
    fetchById: "http://www.dnd5eapi.co/api/spells/{id}",
  },
};
