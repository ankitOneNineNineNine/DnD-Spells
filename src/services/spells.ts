import axios, { AxiosError } from "axios";
import { useQuery } from "react-query";
import { api } from "./api";
import { toastFail } from "./toast";

export interface ISpell {
  index: string;
  name: string;
  url: string;
}
interface ISchool {
  index: string;
  name: string;
  url: string;
}

type IClass = ISpell;
type ISubClass = ISpell;
type IDamageType = ISpell;
export const Componentsenum: { [key: string]: any } = {
  V: "verbal",
  S: "somatic",
  M: "material"
};
export interface ISpellInfo {
  casting_time?: string;
  classes?: IClass[];
  components?: string[];
  concentration?: boolean;
  desc?: string[];
  duration?: string;
  damage?: {
    damage_at_character_level?: { [key: number]: string };
    damage_type?: IDamageType;
  };
  area_of_effect?: {
    size?: number;
    type?: string;
  };
  attack_type?: string;
  higher_level?: string[];
  index?: string;
  level?: 2;
  material?: string;
  name?: string;
  range?: string;
  ritual?: boolean;
  school?: ISchool;
  subclasses?: ISubClass[];
  url?: string;
  _id?: string;
}

export interface IFetchSpellsResponse {
  count: number;
  results: ISpell[];
}

/**
 * Axios Get Call to fetch repos based on search query
 *
 * @returns AxiosResponse<IFetchSpellsResponse>
 */
const fetchSpells = () => () => {
  return axios.get<IFetchSpellsResponse>(api.spells.fetch);
};

/**
 * Axios Get Call to get Spell by Index
 * @param {string} id
 * @returns AxiosResponse<ISpell>
 */
const getSpellByIndex = (id: string) => () => {
  return axios.get<ISpellInfo>(api.spells.fetchById.replace("{id}", id));
};

/**
 * Use Query to Fetch All Spells
 * @export useGetAllSpells
 * @returns UseQueryResult
 */
export const useGetAllSpells = () => {
  return useQuery(api.spells.fetch, fetchSpells(), {
    select: (response) => response.data,
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toastFail(error?.response?.data?.message || "Something Went Wrong");
    }
  });
};

/**
 * Use Query to Fetch Spell By Index
 * @param {string} id
 * @export useGetSpellByIndex
 * @returns UseQueryResult
 */
export const useGetSpellByIndex = (id: string) => {
  return useQuery([api.spells.fetch, id], getSpellByIndex(id), {
    enabled: !!id,
    select: ({ data }) => data,
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toastFail(error?.response?.data?.message || "Something Went Wrong");
    }
  });
};
