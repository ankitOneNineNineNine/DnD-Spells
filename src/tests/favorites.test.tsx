import React from "react";
import { render, screen } from "@testing-library/react";
import { Favorites } from "../pages/Favorites/favorites.pages";
import { ProviderMock } from "../mocks/provider.mock";
import { ISpell } from "../services/spells";

describe("chip", () => {
  function prepare(favorites: ISpell[]) {
    const { container } = render(
      <ProviderMock favorites={favorites}>
        <Favorites closeModal={() => undefined} />
      </ProviderMock>
    );
    return { container };
  }

  it("should match a snapshot", () => {
    const { container } = prepare([]);
    expect(container).toMatchSnapshot();
  });

  it("empty favorite list", () => {
    prepare([]);
    const favoriteList = screen.getByTestId("favoriteList");
    expect(favoriteList.childNodes.length).toBe(0);
  });

  it("non empty favorite list", () => {
    prepare([
      {
        name: "Animal Messenger",
        index: "animal-messenger",
        url: "/api/spells/animal-messenger"
      }
    ]);
    const favoriteList = screen.getByTestId("favoriteList");
    expect(favoriteList.childNodes.length).toBe(1);
  });
});
