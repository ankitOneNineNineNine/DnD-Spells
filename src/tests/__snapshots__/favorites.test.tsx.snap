// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chip should match a snapshot 1`] = `
<div>
  <div
    class="favoriteContainer"
  >
    <h1>
       Favorite Spells 
    </h1>
    <div
      class="closeModal"
      data-testid="close-modal"
    >
      <i
        class="fa fa-close"
      />
    </div>
    <div
      data-testid="favoriteList"
    />
  </div>
</div>
`;
